devel/api.md

```diff
@@ -0,0 +1,1079 @@
+
+In stead of using hard coded string, please use classes and constants in
+`libnmstate.schmea`.
+
+# Limitations
+
+* <https://nmstate.atlassian.net/browse/NMSTATE-235>
+
+  The `libnmstate` does not differentiate `InterfaceState.DOWN` and
+  `InterfaceState.ABSENT`. They both remove the on-disk Network Manager
+  connection profile of specified interface.
+
+* <https://nmstate.atlassian.net/browse/NMSTATE-236>
+
+  All the network state applied by `libnmstate` is preserved after
+  reboot, except for disabling IPv6.
+
+* <https://nmstate.atlassian.net/browse/NMSTATE-115>
+
+  Due to the limitation of NetworkManager, Nmstate does not support disabling
+  DHCPv6 while enabling autoconf yet.
+
+* <https://nmstate.atlassian.net/browse/NMSTATE-212>
+
+  Currently, Nmstate requires the existence of interface section for all the
+  interfaces has route next hop to.
+
+# Network State Main layout
+
+The network state is represented in a dictionary:
+
+```python
+from libnmstate.schema import Interface
+from libnmstate.schema import Route
+from libnmstate.schema import DNS
+
+network_state = {
+    DNS.KEY: {},            # Check section of 'DNS'
+    Route.KEY: {},          # Check section of 'Route'
+    Interface.KEY: [],      # Check section of 'Interface -- Basic'
+}
+```
+
+# Basic Interface
+
+An interface represents a hardware or software network interface.
+Each interface has these basic properties represented as a dictionary.
+The dictionary key strings are stored as constant in `libnmstate.schema`
+module.
+
+## Basic Interface: Example
+
+```python
+{
+    Interface.NAME: 'eth1',
+    Interface.STATE: InterfaceState.UP,
+    Interface.TYPE: InterfaceType.ETHERNET,
+    Interface.IPV4: {
+        InterfaceIPv4.ENABLED: True
+        InterfaceIPv4.DHCP: False,
+        InterfaceIPv4.ADDRESS: [
+            {
+                InterfaceIPv4.ADDRESS_IP: '192.0.2.251',
+                InterfaceIPv4.ADDRESS_PREFIX_LENGTH: 24,
+            }
+        ],
+    },
+    Interface.IPV6: {
+        InterfaceIPv6.ENABLED: True
+        InterfaceIPv6.DHCP: True,
+        InterfaceIPv6.AUTOCONF: True,
+        InterfaceIPv6.AUTO_DNS: True,
+        InterfaceIPv6.AUTO_GATEWAY: True,
+        InterfaceIPv6.AUTO_ROUTES: True,
+        InterfaceIPv6.ADDRESS: [],
+    }
+}
+```
+
+## `Interface.NAME`
+
+The kernel name of the interface.
+
+Type: `string`
+
+## `Interface.TYPE`
+
+The interface type.
+
+Type: `string`
+
+Possible values:
+
+ * `InterfaceType.ETHERNET`
+ * `InterfaceType.BOND`
+ * `InterfaceType.LINUX_BRIDGE`
+ * `InterfaceType.OVS_BRIDGE`
+ * `InterfaceType.OVS_PORT`
+ * `InterfaceType.OVS_INTERFACE`
+ * `InterfaceType.VLAN`
+
+## `Interface.STATE`
+
+The interface state.
+
+Type: `string`
+
+Possible values:
+
+ * `InterfaceState.UP`
+
+   Bring interface link up.
+
+ * `InterfaceState.DOWN`
+
+   Remove interface. Might change to bring interface link down in future
+   release.
+
+ * `InterfaceState.ABSENT`
+
+   Remove interface.
+
```

Comment:
No 'unknown' anymore? I see a 'unknown' value in https://github.com/nmstate/nmstate/blob/master/libnmstate/schemas/operational-state.yaml